packages/fast-usdc/src/exos/liquidity-pool.js

```diff
@@ -0,0 +1,232 @@
+import {
+  AmountMath,
+  AmountShape,
+  PaymentShape,
+  RatioShape,
+} from '@agoric/ertp';
+import {
+  makeRecorderTopic,
+  TopicsRecordShape,
+} from '@agoric/zoe/src/contractSupport/topics.js';
+import { depositToSeat } from '@agoric/zoe/src/contractSupport/zoeHelpers.js';
+import { SeatShape } from '@agoric/zoe/src/typeGuards.js';
+import { M } from '@endo/patterns';
+import { Fail, q } from '@endo/errors';
+import {
+  depositCalc,
+  makeParity,
+  withdrawCalc,
+  withFees,
+} from '../pool-share-math.js';
+import { makeProposalShapes } from '../type-guards.js';
+
+/**
+ * @import {Zone} from '@agoric/zone';
+ * @import {Remote, TypedPattern} from '@agoric/internal'
+ * @import {StorageNode} from '@agoric/internal/src/lib-chainStorage.js'
+ * @import {MakeRecorderKit, RecorderKit} from '@agoric/zoe/src/contractSupport/recorder.js'
+ * @import {USDCProposalShapes, ShareWorth} from '../pool-share-math.js'
+ */
+
+const { add, isEqual } = AmountMath;
+
+/** @param {Brand} brand */
+const makeDust = brand => AmountMath.make(brand, 1n);
+
+/**
+ *
+ * @param {ZCFSeat} poolSeat
+ * @param {ShareWorth} shareWorth
+ * @param {Brand} USDC
+ */
+const checkPoolBalance = (poolSeat, shareWorth, USDC) => {
+  const available = poolSeat.getAmountAllocated('USDC', USDC);
+  const dust = makeDust(USDC);
+  isEqual(add(available, dust), shareWorth.numerator) ||
+    Fail`🚨 pool balance ${q(available)} inconsistent with shareWorth ${q(shareWorth)}`;
+};
```

Comment:
1. Consider `assertPoolBalance`
2. Can you talk more about what this invariant check is doing? How will it work when an advance is in flight?

Reply:
> Consider assertPoolBalance

`assertFoo(x)` usually means: assert that `x` is a `Foo`.

> Can you talk more about what this invariant check is doing?

The `atomicRearrange`s that follow the call to `depositCalc` / `withdrawCalc` assume that the pool balance represented by the `poolSeat` USDC allocation is the same as the pool balance represented by the numerator of `shareWorth`. When I first added this check, my tests failed; they differ by the dust used to make the initial shareworth denominator non-zero.

The fact that this answer didn't come in the form of a pointer to existing docs says I should add more.

> How will it work when an advance is in flight?

In flight in what sense? Its only use is in normal straight-line synchronous offer handlers: gather all the relevant info in one place and compute the outcome, synchronously. Things happening outside this vat may happen causally before or after, or in an incomparable causal order. In any case, the local computation does what it does.